server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java

```diff
@@ -521,60 +517,25 @@ private boolean checkForOrderDataChanges(Pool pool, Pool existingPool) {
         return orderDataChanged;
     }
 
-    private Set<Product> getExpectedProvidedProducts(Pool pool, boolean useDerived) {
-        Set<Product> incomingProvided = new HashSet<>();
-        /**
-         * It is necessary to use getters for provided products here, because the pool
-         * is fabricated from subscrfiption (using CandlepinPoolManager.convertToMasterPool
-         * It is not an actual pool that would be stored in the DB.
-         */
-        Set<Product> source = useDerived ? pool.getDerivedProvidedProducts() : pool.getProvidedProducts();
-
-        if (source != null && !source.isEmpty()) {
-            incomingProvided.addAll(source);
-        }
-
-        return incomingProvided;
-    }
-
-    private boolean changedProductsInSet(Set<Product> products, Map<String, Product> changedProducts) {
-
-        if (products != null && changedProducts != null) {
-            for (Product product : products) {
-                if (product != null && changedProducts.get(product.getId()) != null) {
-                    return true;
-                }
-            }
-
-        }
-        return false;
-    }
-
-    private boolean checkForChangedProducts(Product incomingProduct, Set<Product> incomingProvided,
-        Pool existingPool, Map<String, Product> changedProducts) {
+    private boolean checkForChangedProducts(Product incomingProduct, Pool existingPool, Map<String,
+        Product> changedProducts) {
 
         Product existingProduct = existingPool.getProduct();
-        Set<Product> currentProvided = productCurator.getPoolProvidedProductsCached(existingPool);
         String pid = existingProduct.getId();
 
         // TODO: ideally we would differentiate between these different product changes
         // a little, but in the end it probably doesn't matter:
-        boolean productsChanged =
-            (pid != null && !pid.equals(incomingProduct.getId())) ||
-            !currentProvided.equals(incomingProvided);
+        boolean productsChanged = pid != null && !pid.equals(incomingProduct.getId());
```

Comment:
The `changedProducts` Map is still using productIds for comparison, not sure if we can remove it.
If so we can remove it, this whole bit can be refactored along with `checkForChangedDerivedProducts` method.

Reply:
We probably don't need both anymore (if a derived product changed, the product itself changed and should show up in changedProducts); hence why I said the whole flow could use a review. The original intent is slowly getting lost, and it would do us some good to re-evaluate and refactor this whole area.

Reply:
Earlier we discussed that product (both mkt and eng product) member properties are equally 
essential to detect if a product has changed or not. The `changedProducts` map contains 
the net sum of any of the properties of a product that have changed or not. This map is populated inside ProductManager while products are getting imported.
Now in this current method we cant just only rely on the productID or Uuid to mark the product has changed from state present in pool (product in existing pool), therefore changeProduct map becomes equally important for this. 
Any thoughts on this?

Reply:
One oddity I just realized with this process, is we're probably breaking it when we do the reference updating via SQL at the end of importProduct, even today in current master.

Hrmm...

Reply:
That was a good idea to use Products entity version to detect if the incoming product (present in Pool) has changed from existing products (present in existing pools) in this particular code block.

Based on preliminary observation there is one issue, when we trace the refresh flow, we first import products, if there is any changes we update (or create a new one) the product and finally persist it in DB. 

In next couple of step we get all the subscription pools - 
`CandlepinPoolManger.java` Method - `refreshPoolsForMasterPool`
```
subscriptionPools = this.poolCurator.getPoolsBySubscriptionId(pool.getSubscriptionId()).list();
```
When we make this DB call, we get all the updated products which was already persisted in DB during Product import call (with updated entity version of course) inside the subscription pools (aka existing pools).

This is why I was wondering why the entity version of incoming product and product inside existing pools are same, because they are same product. This could be one of the reason why we need a `changedProduct` map.

Reply:
@Ceiu  Any thoughts on above comment ?

Reply:
I wouldn't be surprised if that's exactly what's happening. It probably comes down to whether or not a given pool is already in the session cache before the import call is made.

We can leave it as-is for now, and revisit it once we have time to revisit the pool side of refresh.